geopy/geocoders/here.py

```diff
@@ -0,0 +1,281 @@
+from geopy.exc import (
+    GeocoderAuthenticationFailure,
+    GeocoderInsufficientPrivileges,
+    GeocoderQuotaExceeded,
+    GeocoderServiceError,
+    GeocoderUnavailable,
+)
+from geopy.geocoders.base import DEFAULT_SENTINEL, Geocoder
+from geopy.location import Location
+from geopy.util import join_filter, logger
+
+__all__ = ("Here", )
+
+
+class Here(Geocoder):
+    """Geocoder using the HERE Geocoder API.
+
+    Documentation at:
+        https://developer.here.com/documentation/geocoder/
+    """
+
+    structured_query_params = {
+        'city',
+        'county',
+        'district',
+        'country',
+        'state',
+        'street',
+        'housenumber',
+        'postalcode',
+    }
+
+    def __init__(
+            self,
+            app_id,
+            app_code,
+            format_string=None,
+            scheme=None,
+            timeout=DEFAULT_SENTINEL,
+            proxies=DEFAULT_SENTINEL,
+            user_agent=None,
+            ssl_context=DEFAULT_SENTINEL,
+    ):
+        """
+
+        :param str app_id: Should be a valid HERE Maps APP ID
+            See https://developer.here.com/authenticationpage.
+
+        :param str app_code: Should be a valid HERE Maps APP CODE
+            See https://developer.here.com/authenticationpage.
+
+        :param str format_string:
+            See :attr:`geopy.geocoders.options.default_format_string`.
+
+        :param str scheme:
+            See :attr:`geopy.geocoders.options.default_scheme`.
+
+        :param int timeout:
+            See :attr:`geopy.geocoders.options.default_timeout`.
+
+        :param dict proxies:
+            See :attr:`geopy.geocoders.options.default_proxies`.
+
+        :param str user_agent:
+            See :attr:`geopy.geocoders.options.default_user_agent`.
+
+        :type ssl_context: :class:`ssl.SSLContext`
+        :param ssl_context:
+            See :attr:`geopy.geocoders.options.default_ssl_context`.
+        """
+        super(Here, self).__init__(
+            format_string=format_string,
+            scheme=scheme,
+            timeout=timeout,
+            proxies=proxies,
+            user_agent=user_agent,
+            ssl_context=ssl_context,
+        )
+        self.app_id = app_id
+        self.app_code = app_code
+        self.api = "%s://geocoder.api.here.com/6.2/geocode.json" % self.scheme
+        self.reverse_api = "%s://reverse.geocoder.api.here.com/6.2/reversegeocode.json" \
+            % self.scheme
+        # FIXME(?): add multi-reverse geocode
+        # FIXME(?): add landmark geocode
+
+    def geocode(
+            self,
+            query,
+            exactly_one=True,
+            language=None,
+            additional_data=False,
+            other=None,
+            timeout=DEFAULT_SENTINEL
+    ):
+        """
+        Return a location point by address.
+
+        Documentation at:
+            https://developer.here.com/documentation/geocoder/topics/resource-geocode.html
```

Comment:
`geocode`/`reverse` methods usually don't contain links to the relevant API docs, given that there's already a link in the class's docstring. Also the links tend to change, so having a single link is easier for maintenance and is less cluttering.

What do you think? I guess it was helpful to have these links around with the `other` kwarg, but without it they seem to be less useful.

Reply:
It's a matter of taste, indeed. I'm fine with keeping them if you think they're helpful there.